src/main/java/com/nvidia/spark/rapids/jni/GpuTimeZoneDB.java

```diff
@@ -26,72 +33,171 @@
 import java.util.List;
 import java.util.Map;
 import java.util.TimeZone;
-import java.util.concurrent.*;
-
-import ai.rapids.cudf.ColumnVector;
-import ai.rapids.cudf.DType;
-import ai.rapids.cudf.HostColumnVector;
-import ai.rapids.cudf.Table;
+import java.util.concurrent.Executors;
 
+/**
+ * Gpu time zone utility.
+ * Provides two kinds of APIs
+ *  - Time zone transitions cache APIs
+ *      `cacheDatabaseAsync`, `cacheDatabase` and `shutdown` are synchronized.
+ *      When cacheDatabaseAsync is running, the `shutdown` and `cacheDatabase` will wait;
+ *      These APIs guarantee only one thread is loading transitions cache,
+ *      And guarantee loading cache only occurs one time.
+ *  - Rebase time zone APIs
+ *    fromTimestampToUtcTimestamp, fromUtcTimestampToTimestamp ...
+ */
 public class GpuTimeZoneDB {
-
-  public static final int TIMEOUT_SECS = 300;
-
+  private static final Logger log = LoggerFactory.getLogger(GpuTimeZoneDB.class);
 
   // For the timezone database, we store the transitions in a ColumnVector that is a list of 
   // structs. The type of this column vector is:
   //   LIST<STRUCT<utcInstant: int64, localInstant: int64, offset: int32>>
-  private CompletableFuture<Map<String, Integer>> zoneIdToTableFuture;
-  private CompletableFuture<HostColumnVector> fixedTransitionsFuture;
+  private Map<String, Integer> zoneIdToTable;
 
-  private boolean closed = false;
+  // use this reference to indicate if time zone cache is initialized.
+  private HostColumnVector fixedTransitions;
 
-  GpuTimeZoneDB() {
-    zoneIdToTableFuture = new CompletableFuture<>();
-    fixedTransitionsFuture = new CompletableFuture<>();
+  // Guarantee singleton instance
+  private GpuTimeZoneDB() {
   }
-  
-  private static GpuTimeZoneDB instance = new GpuTimeZoneDB();
-  // This method is default visibility for testing purposes only. The instance will be never be exposed publicly
-  // for this class.
+
+  // singleton instance
+  private static final GpuTimeZoneDB instance = new GpuTimeZoneDB();
+
+  // This method is default visibility for testing purposes only.
+  // The instance will be never be exposed publicly for this class.
   static GpuTimeZoneDB getInstance() {
     return instance;
   }
-  
+
+  static class LoadingLock {
+    Boolean isLoading = false;
+
+    // record whether a shutdown is called ever.
+    // if `isCloseCalledEver` is true, then the following loading should be skipped.
+    Boolean isShutdownCalledEver = false;
+  }
```

Comment:
I'm not sure we need a separate lock class for this rather than locking the whole database via synchronized methods as was done previously.  We just need to lock the GpuTimeZoneDB instance when manipulating the boolean flags which can be direct members of the database. cacheDatabaseAsync, cacheDatabaseImpl and shutdown can just be synchronized methods, which removes the need for most of them to wait on threads -- the wait will be implicit in obtaining the database lock for the synchronized method. The only wait we need is in the shutdown method, and only when isLoading is true.

By leaving the methods synchronized rather than having a separate lock class, it's a little easier to reason about the synchronization safety because we don't have to keep track of whether we're holding LoadingLock at the right times.